### PR TITLE
bpo-30565: Add PYTHONCOERCECLOCALE=warn runtime flag

### PR DESCRIPTION
**Pull request by ncoghlan**
- legacy locale warnings are now silent by default
- build time configuration setting is removed
- set PYTHONCOERCECLOCALE=warn to enable them
- updates test cases accordingly

### CODE REVIEW

**Comment by ncoghlan** [comment]
Marking as WIP, as we shouldn't merge this until after we get custom build results back from the BuildBot fleet.

**Comment by ncoghlan** [comment]
Custom build link for one of the currently failing buildbots that this change is expected to fix: http://buildbot.python.org/all/builders/AMD64%20FreeBSD%20CURRENT%20Debug%20custom/builds/8

**Review by vstinner — APPROVED** [review]
LGTM but it seems that tests fail :-)

**Comment by ncoghlan** [comment]
Let's try that again with a test case setup that doesn't immediately fail when no coercion targets are available: http://buildbot.python.org/all/builders/AMD64%20FreeBSD%20CURRENT%20Debug%20custom/builds/9

**Comment by ncoghlan** [comment]
The previous custom build succeeded (with warnings), but I realised that there were several test cases that could actually be run even when no coercion target locales are available, as long as their expectations were adjusted accordingly.

So I've updated the PR to do that, and initiated another custom build run: http://buildbot.python.org/all/builders/AMD64%20FreeBSD%20CURRENT%20Debug%20custom/builds/10

**Comment by ncoghlan** [comment]
One of the checks for "No coercion target locales available" was incorrect, so: http://buildbot.python.org/all/builders/AMD64%20FreeBSD%20CURRENT%20Debug%20custom/builds/11

**Comment by ncoghlan** [comment]
Making a note so *I* don't forget: the other reason this is marked WIP is because I haven't updated the What's New entry and the `PYTHONCOERCECLOCALE` documentation yet.

**Comment by ncoghlan** [comment]
Ah, looks like the updated logic still isn't quite right - for the cases where the tests currently expect a warning (regardless of which warning they expect), they need to be updated to expect to see the legacy locale warning when there aren't any coercion target locales available.

**Comment by ncoghlan** [comment]
Skipping UTF-8 as a coercion target for now has got the FreeBSD buildbots back to their pre-PEP-538 state: http://buildbot.python.org/all/builders/AMD64%20FreeBSD%20CURRENT%20Debug%20custom/builds/13/steps/test/logs/stdio

I've also made the required docs and help changes, so assuming CI completes successfully, I'll merge this, and we can work out the details of the *actual* expected behaviour on \*BSD systems (including Mac OS X) before re-enabling PEP 538 there.

**Pull request merged**
